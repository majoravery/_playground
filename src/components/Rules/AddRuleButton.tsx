import React, { ComponentProps } from "react";
import { Button } from "@deliveryhero/armor";
import { PlusIcon } from "@deliveryhero/armor-icons";

interface AddRuleButtonProps extends ComponentProps<typeof Button> {
  title?: string;
}

const AddRuleButton: React.FC<AddRuleButtonProps> = ({
  title = "Add rule",
  ...props
}) => (
  <Button marginTop={4} secondary small {...props}>
    <PlusIcon /> {title}
  </Button>
);

export default AddRuleButton;
