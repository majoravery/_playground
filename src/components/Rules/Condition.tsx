/* eslint-disable react/prop-types */
import React from "react";
import { ICondition } from "./types";
import ViewCondition from "./ViewCondition";
import EditCondition from "./EditCondition";

interface Props extends ICondition {
  first?: boolean;
}

const Condition: React.FC<Props> = ({
  conditionKey,
  conditionIndex,
  logicalOperator,
  onLogicalOperatorChange,
  logicalOperatorReadOnly,
  comparator,
  values,
  title,
  mode,
  conditionReadOnly = false,
  first,
  onEditCondition,
  onDeleteCondition,
  onViewCondition,
  disableViewCondition,
  disableEditCondition,
  disableDeleteCondition,
  ValuesComponent,
  valuesComponentProps,
  children,
}) => {
  return (
    <>
      {mode === "view" ? (
        <ViewCondition
          comparator={comparator}
          conditionIndex={conditionIndex}
          conditionKey={conditionKey}
          conditionReadOnly={conditionReadOnly}
          disableDeleteCondition={disableDeleteCondition}
          disableEditCondition={disableEditCondition}
          disableViewCondition={disableViewCondition}
          first={first}
          logicalOperator={logicalOperator}
          logicalOperatorReadOnly={logicalOperatorReadOnly}
          onDeleteCondition={onDeleteCondition}
          onEditCondition={onEditCondition}
          onLogicalOperatorChange={onLogicalOperatorChange}
          onViewCondition={onViewCondition}
          title={title}
          values={values}
          ValuesComponent={ValuesComponent}
          valuesComponentProps={valuesComponentProps}
        />
      ) : (
        <EditCondition
          conditionIndex={conditionIndex}
          conditionKey={conditionKey}
          values={values}
        />
      )}
      {children}
    </>
  );
};

export default Condition;
