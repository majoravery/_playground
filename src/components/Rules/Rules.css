.rule__container {
    background: #fafafd;
    border-radius: 4px;
    padding: 24px;
}

.rule__add-condition-button {
    width: 148px;
    font-weight: 500;
}

.rule__header {
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: space-between;
    margin-bottom: 28px;
    height: 32px;
}

.rule__header>* {
    margin: 0;
}

.rule__conditions-container {
    margin-bottom: 24px;
}

.rule__conditions-container>*+* {
    margin-top: 16px;
}

.rule__add-button {
    width: 100%;
    height: 64px;
    border-radius: 4px;
    font-weight: 500;
}

.rule-label__container {
    box-sizing: border-box;
    display: flex;
    height: 40px;
    background: #efeef6;
    border-radius: 4px;
    padding: 12px 16px;
    align-items: center;
}

.view-condition__container {
    display: flex;
    flex-direction: row;
}

.view-condition__container>*+* {
    margin-left: 8px;
}

.view-condition__container .dropdown-toggle {
    height: 40px;
    width: 112px;
    text-align: left;
    display: flex;
    align-items: center;
}

.view-condition__container .dropdown-toggle span {
    flex: 1;
}

.view-condition__container .dropdown-menu {
    min-width: 112px;
}

.values__container {
    box-sizing: border-box;
    min-height: 40px;
    background: white;
    border: 1px solid #d8d8d8;
    border-radius: 4px;
    padding: 3px 7px;
    display: flex;
    align-items: center;
    flex: 1;
}

.values__content-container {
    flex: 1;
}

.values__button-container {
    height: 100%;
}

.edit-condition__text--title {
    height: 20px;
    color: #5a59a9;
    font-size: 14px;
    font-weight: 500;
    letter-spacing: 0.3px;
    line-height: 20px;
    margin-bottom: 8px;
}

.edit-condition__text--subtitle {
    height: 20px;
    color: #2d2d36;
    font-size: 14px;
    font-weight: normal;
    letter-spacing: 0.3px;
    line-height: 20px;
    margin-bottom: 8px;
}

.edit-condition__container {
    min-height: 505px;
    background: white;
    border: 1px solid #d8d8d8;
    border-radius: 4px;
    display: flex;
    flex-direction: row;
}

.edit-condition__sidebar {
    box-shadow: none !important;
    border-right: 1px solid #d8d8d8;
}

.edit-condition__content {
    display: flex;
    flex-direction: column;
    flex: 1;
    padding: 24px;
}

.edit-condition__custom-content {
    display: flex;
    flex-wrap: wrap;
    flex: 1;
}

.rules__container>*+* {
    margin-top: 24px;
}