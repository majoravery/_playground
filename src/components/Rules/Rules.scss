.rule {

  &__container {
    background: rgb(250, 250, 253);
    border-radius: 4px;
    padding: 24px;
  }

  &__add-condition-button {
    width: 148px;
    font-weight: 500;
  }

  &__header {
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: space-between;
    margin-bottom: 28px;
    height: 32px;

    > * {
      margin: 0;
    }
  }

  &__conditions-container {
    margin-bottom: 24px;

    > {

      * + * {
        margin-top: 16px;
      }
    }
  }

  &__add-button {
    width: 100%;
    height: 64px;
    border-radius: 4px;
    font-weight: 500;
  }
}

.rule-label {

  &__container {
    display: flex;
    height: 40px;
    background: rgb(239, 238, 246);
    border-radius: 4px;
    padding: 12px 16px;
    align-items: center;
  }
}

.view-condition {

  &__container {
    display: flex;
    flex-direction: row;

    > {

      * + * {
        margin-left: 8px;
      }
    }

    .dropdown-toggle {
      height: 40px;
      width: 112px;
      text-align: left;
      display: flex;
      align-items: center;

      span {
        flex: 1;
      }
    }

    .dropdown-menu {
      min-width: 112px;
    }
  }
}

.values {

  &__container {
    min-height: 40px;
    background: rgb(255, 255, 255);
    border: 1px solid rgb(216, 216, 216);
    border-radius: 4px;
    padding: 3px 7px;
    display: flex;
    align-items: center;
    flex: 1;
  }

  &__content-container {
    flex: 1;
  }

  &__button-container {
    height: 100%;
  }
}

.edit-condition {

  &__text {

    &--title {
      height: 20px;
      color: rgb(90, 89, 169);
      font-size: 14px;
      font-weight: 500;
      letter-spacing: 0.3px;
      line-height: 20px;
      margin-bottom: 8px;
    }

    &--subtitle {
      height: 20px;
      color: rgb(45, 45, 54);
      font-size: 14px;
      font-weight: normal;
      letter-spacing: 0.3px;
      line-height: 20px;
      margin-bottom: 8px;
    }
  }

  &__container {
    min-height: 505px;
    background: rgb(255, 255, 255);
    border: 1px solid rgb(216, 216, 216);
    border-radius: 4px;
    display: flex;
    flex-direction: row;
  }

  &__sidebar {
    box-shadow: none !important;
    border-right: 1px solid rgb(216, 216, 216);
  }

  &__content {
    display: flex;
    flex-direction: column;
    flex: 1;
    padding: 24px;
  }

  &__custom-content {
    flex: 1;
  }
}

.rules {

  &__container {

    > {

      * + * {
        margin-top: 24px;
      }
    }
  }
}
