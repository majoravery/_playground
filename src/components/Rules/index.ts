import {
  IRules as IRulesAlias,
  IRule as IRuleAlias,
  TRule as TRuleAlias,
  TOption as TOptionAlias,
  TCondition as TConditionAlias,
  ICondition as IConditionAlias,
  IRuleState as IRuleStateAlias,
  UserComponentBaseProps as UserComponentBasePropsAlias,
  Key as KeyAlias,
  IRuleCustomRenderProps as IRuleCustomRenderPropsAlias,
} from './types';

export { default as Rules } from './Rules';
export { default as Rule } from './Rule';
export { default as Values } from './Values';
export { RuleStateProvider } from './hooks/useRuleState';
export { default as AddRuleButton } from './AddRuleButton';
export { default as RuleContainer } from './RuleContainer';
export { default as Condition } from './Condition';
export { default as EditCondition } from './EditCondition';

export type IRules = IRulesAlias;
export type IRule = IRuleAlias;
export type TRule = TRuleAlias;
export type TOption = TOptionAlias;
export type TCondition = TConditionAlias;
export type ICondition = IConditionAlias;
export type IRuleState = IRuleStateAlias;
export type UserComponentBaseProps = UserComponentBasePropsAlias;
export type Key = KeyAlias;
export type IRuleCustomRenderProps = IRuleCustomRenderPropsAlias;
