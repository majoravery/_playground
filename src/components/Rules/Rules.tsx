import React from "react";
import { TRule, TOption, TCondition, IRuleState, IRules } from "./types";
import Rule from "./Rule";
import AddRuleButton from "./AddRuleButton";
import { produce } from "immer";

type TSetOptions = (
  setRules: IRules["setRules"],
  i: number
) => (options: TOption[] | ((prev: TOption[]) => TOption[])) => void;

type TSetConditions = (
  setRules: IRules["setRules"],
  i: number
) => (
  conditions: TCondition[] | ((prev: TCondition[]) => TCondition[])
) => void;

type TSetAddNewMode = (
  setRules: IRules["setRules"],
  i: number
) => (mode: boolean | ((prev: boolean) => boolean)) => void;

export const setOptions: TSetOptions = (setRules, i) => (options) => {
  setRules((oldRules: TRule[]) => {
    let newOptions: TOption[];
    if (typeof options === "function") {
      newOptions = options(oldRules[i].options);
    } else {
      newOptions = options;
    }

    return produce(oldRules, (draft) => {
      draft[i].options = newOptions;
    });
  });
};

export const setConditions: TSetConditions = (setRules, i) => (conditions) => {
  setRules((oldRules: TRule[]) => {
    let newConditions: TCondition[];
    if (typeof conditions === "function") {
      newConditions = conditions(oldRules[i].conditions);
    } else {
      newConditions = conditions;
    }

    return produce(oldRules, (draft) => {
      draft[i].conditions = newConditions;
    });
  });
};

export const setAddNewMode: TSetAddNewMode = (setRules, i) => (mode) => {
  setRules((oldRules) => {
    let newMode: boolean;
    if (typeof mode === "function") {
      newMode = mode(oldRules[i].addNewMode);
    } else {
      newMode = mode;
    }

    return produce(oldRules, (draft) => {
      draft[i].addNewMode = newMode;
    });
  });
};

const Rules: React.FC<IRules> = ({
  rules,
  setRules,
  enableAdd = false,
  addButtonTitle,
  onAddRule,
  onDeleteRule,
  readonly = false,
}) => {
  // gets rule props from user, and will not add rule if onAddRule does not return object
  const handleAddRule = () => {
    const newRuleProps = onAddRule();
    if (typeof newRuleProps !== "object") return;
    const {
      conditions = [],
      addNewMode = false,
      readonly = false,
      ...others
    } = newRuleProps;
    const newRule: TRule = {
      conditions,
      addNewMode,
      readonly,
      ...others,
    };

    setRules((oldRules) =>
      produce(oldRules, (draft) => {
        draft.push(newRule);
      })
    );
  };

  // calls user's onDeleteRule, and only continues if it returns true
  const handleDeleteRule = (deleteProps: IRuleState) => {
    if (onDeleteRule !== undefined) {
      if (!onDeleteRule(deleteProps)) return;
    }

    setRules((oldRules) =>
      produce(oldRules, (draft) => {
        const index = draft.findIndex(
          (rule) => rule.ruleKey === deleteProps.ruleKey
        );
        if (index !== -1) draft.splice(index, 1);
      })
    );
  };

  return (
    <div className="rules__container" data-testid="rules--container">
      {rules.map((rule, i) => (
        <Rule
          addNewMode={rule.addNewMode}
          conditions={rule.conditions}
          disableDeleteRule={rule.disableDeleteRule}
          enableAddCondition={rule.enableAddCondition}
          hideDeleteRule={rule.hideDeleteRule}
          key={rule.ruleKey}
          onAddCondition={rule.onAddCondition}
          onDeleteRule={handleDeleteRule}
          options={rule.options}
          optionsSidebarTitle={rule.optionsSidebarTitle}
          readonly={rule.readonly || readonly}
          render={rule.render}
          ruleKey={rule.ruleKey}
          setAddNewMode={setAddNewMode(setRules, i)}
          setConditions={setConditions(setRules, i)}
          setOptions={setOptions(setRules, i)}
          title={rule.title}
        />
      ))}
      {enableAdd && !readonly && (
        <AddRuleButton onClick={handleAddRule} title={addButtonTitle} />
      )}
    </div>
  );
};

export default Rules;
