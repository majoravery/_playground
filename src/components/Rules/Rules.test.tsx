import React, { useState } from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Rules, TOption, TCondition, TRule, UserComponentBaseProps } from './index';
import { IRules } from './types';
import { produce } from 'immer';
import { setOptions, setConditions, setAddNewMode } from './Rules';

const TestRules: React.FC<Partial<IRules>> = props => {
  const [rules, setRules] = useState<TRule[]>(props.rules || []);
  const handleAddRule = props.onAddRule || jest.fn();
  const handleDeleteRule = props.onDeleteRule || jest.fn();
  const enableAdd = props.enableAdd === undefined ? true : props.enableAdd;
  const readonly = props.readonly;

  return (
    <Rules
      rules={rules}
      setRules={setRules}
      onAddRule={handleAddRule}
      onDeleteRule={handleDeleteRule}
      enableAdd={enableAdd}
      readonly={readonly}
    />
  );
};

let conditionCount = 0;
const type = 'type1';
const newTitle = 'New Title';
const cancelConditionButtonText = 'Cancel Condition';
const applyConditionButtonText = 'Apply Condition';
export const EditComponent: React.FC<UserComponentBaseProps> = ({
  conditionKey,
  setConditions,
  setAddNewMode,
  options,
  setOptions,
  readonly,
}) => {
  const handleCancel = () => {
    setConditions(oldConditions =>
      produce(oldConditions, draft => {
        if (conditionKey === undefined) {
          // new condition
          setAddNewMode(false);
          return;
        }
        const selectedCondition = draft.find(c => c.conditionKey === conditionKey);
        selectedCondition!.mode = 'view';
      })
    );
  };

  const handleApply = () => {
    if (conditionKey === undefined) {
      // new condition
      setConditions(oldConditions =>
        produce(oldConditions, draft => {
          draft.push({
            conditionKey: `key-${++conditionCount}`,
            type: options[0].type,
            logicalOperator: 'and',
            comparator: 'is',
            values: {
              list: ['neu1', 'new2'],
            },
            mode: 'view',
          });
        })
      );
      setAddNewMode(false);

      // test set options
      setOptions(oldOptions =>
        produce(oldOptions, draft => {
          const selectedOptions = draft.find(o => o.type === type);
          selectedOptions!.title = 'New Title';
        })
      );

      return;
    }

    setConditions(oldConditions =>
      produce(oldConditions, draft => {
        const selectedCondition = draft.find(c => c.conditionKey === conditionKey);
        selectedCondition!.mode = 'view';
      })
    );
  };

  return (
    <div>
      {!readonly && (
        <>
          <button onClick={handleCancel}>{cancelConditionButtonText}</button>
          <button onClick={handleApply}>{applyConditionButtonText}</button>
        </>
      )}
    </div>
  );
};

const ValuesComponent: React.FC<UserComponentBaseProps> = ({ values }) => {
  let listValues: string[] = [];
  if (values && values.list && Array.isArray(values.list)) {
    listValues = values?.list || [];
  }

  return (
    <div style={{ marginTop: -1, marginBottom: -1 }}>
      {listValues.map((v, i) => (
        <div key={i}>{v}</div>
      ))}
    </div>
  );
};

describe('Rules component', () => {
  it('should render', () => {
    expect(
      <Rules
        rules={[]}
        onAddRule={jest.fn()}
        setRules={jest.fn()}
        readonly={false}
        onDeleteRule={jest.fn()}
        enableAdd={false}
      />
    ).not.toBeNull();
  });

  it('should match snapshot when empty', () => {
    const { container, rerender } = render(<TestRules enableAdd={false} />);
    expect(container).toMatchSnapshot(); // disabled add
    rerender(<TestRules enableAdd={true} />);
    expect(container).toMatchSnapshot(); // enabled add
  });

  it('should be able to add and remove rules and conditions', () => {
    let ruleCount = 0;
    const options: TOption[] = [
      {
        type,
        title: 'Title1',
        contentTitle: 'Title1',
        contentSubtitle: 'Subtitle1',
        EditComponent,
        editComponentProps: {
          foo: 'default props',
        },
        disabled: false,
        ValuesComponent: ValuesComponent,
      },
    ];

    const handleOnAddRule = () => {
      ruleCount += 1;
      return {
        ruleKey: `rk-${ruleCount}`,
        title: `title-${ruleCount}`,
        options,
      };
    };

    const handleDeleteRule = jest.fn(() => true);

    const maxRuleCount = 10;

    const { container, queryByText, queryAllByText, rerender } = render(
      <TestRules rules={[]} onAddRule={handleOnAddRule} onDeleteRule={handleDeleteRule} />
    );
    const emptyRules = container.querySelector('.rule__container');
    expect(emptyRules).toBeNull();
    const addNewRuleButton = queryByText('Add rule');
    expect(addNewRuleButton).not.toBeNull();

    // Add Rules
    for (let i = 0; i < maxRuleCount; i++) {
      fireEvent.click(addNewRuleButton!);
      const ruleCount = container.querySelectorAll('.rule__container').length;
      expect(ruleCount).toEqual(i + 1);
    }

    // Add Conditions
    const maxConditionCount = 3;
    let conditionContainers = container.querySelectorAll('.view-condition__container');
    expect(conditionContainers.length).toEqual(0);

    for (let i = 0; i < maxConditionCount; i++) {
      const addConditionButtons = queryAllByText('Add Condition');
      expect(addConditionButtons[0]).not.toBeNull();
      fireEvent.click(addConditionButtons[0]);
      const applyButton = queryByText(applyConditionButtonText);
      expect(applyButton).not.toBeNull();
      fireEvent.click(applyButton!);
      conditionContainers = container.querySelectorAll('.view-condition__container');
      expect(conditionContainers.length).toEqual(i + 1);
    }

    // test component should update rule title on apply
    expect(queryAllByText(newTitle).length > 0).toBe(true);

    expect(conditionContainers).toMatchSnapshot();

    // should not allow edits when read-only
    rerender(<TestRules rules={[]} onAddRule={handleOnAddRule} onDeleteRule={handleDeleteRule} readonly />);
    expect(queryByText('Add Condition')).toBeNull();
    expect(queryByText('Add rule')).toBeNull();
    expect(queryByText('title-1')!.parentElement!.querySelector('button')).toBeNull();
    const readOnlyCondition = queryAllByText(newTitle)[0].parentElement;
    expect(readOnlyCondition).toMatchSnapshot();

    // set back to non-readonly (editable)
    rerender(<TestRules rules={[]} onAddRule={handleOnAddRule} onDeleteRule={handleDeleteRule} />);

    // Remove Rules
    for (let i = 0; i < maxRuleCount; i++) {
      const title = `title-${i + 1}`;
      const titleElement = queryByText(title);
      expect(titleElement).not.toBeNull();
      const deleteRuleButton = titleElement!.parentElement!.querySelector('button');
      fireEvent.click(deleteRuleButton!);
      expect(handleDeleteRule).toHaveBeenCalled();
      const noTitleElement = queryByText(title);
      expect(noTitleElement).toBeNull();
    }
  });
});

describe('setOptions', () => {
  it('should set options when type is value', () => {
    let newRules: TRule[] = [];
    const newOptionsValue: TOption[] = [];
    const oldRules: Partial<TRule>[] = [
      {
        options: [],
      },
    ];
    const setRules = (newValue: (oldRules: TRule[]) => TRule[]) => {
      newRules = newValue(oldRules as TRule[]);
    };
    const setOptionsFn = setOptions(setRules as IRules['setRules'], 0);
    setOptionsFn(newOptionsValue);
    expect(newRules).not.toBeUndefined();
    expect(newRules[0].options === newOptionsValue).toBeTruthy();
  });

  it('should set options when type is function', () => {
    let newRules: TRule[] = [];
    const newOptionsValue: TOption[] = [];
    const newOptionsFn = () => newOptionsValue;
    const oldRules: Partial<TRule>[] = [
      {
        options: [],
      },
    ];
    const setRules = (newValue: (oldRules: TRule[]) => TRule[]) => {
      newRules = newValue(oldRules as TRule[]);
    };
    const setOptionsFn = setOptions(setRules as IRules['setRules'], 0);
    setOptionsFn(newOptionsFn);
    expect(newRules).not.toBeUndefined();
    expect(newRules[0].options === newOptionsValue).toBeTruthy();
  });
});

describe('setConditions', () => {
  it('should set conditions when type is value', () => {
    let newRules: TRule[] = [];
    const newConditionsValue: TCondition[] = [];
    const oldRules: Partial<TRule>[] = [
      {
        conditions: [],
      },
    ];
    const setRules = (newValue: (oldRules: TRule[]) => TRule[]) => {
      newRules = newValue(oldRules as TRule[]);
    };
    const setConditionsFn = setConditions(setRules as IRules['setRules'], 0);
    setConditionsFn(newConditionsValue);
    expect(newRules).not.toBeUndefined();
    expect(newRules[0].conditions === newConditionsValue).toBeTruthy();
  });

  it('should set conditions when type is function', () => {
    let newRules: TRule[] = [];
    const newConditionsValue: TCondition[] = [];
    const newConditionsValueFn = () => newConditionsValue;
    const oldRules: Partial<TRule>[] = [
      {
        conditions: [],
      },
    ];
    const setRules = (newValue: (oldRules: TRule[]) => TRule[]) => {
      newRules = newValue(oldRules as TRule[]);
    };
    const setConditionsFn = setConditions(setRules as IRules['setRules'], 0);
    setConditionsFn(newConditionsValueFn);
    expect(newRules).not.toBeUndefined();
    expect(newRules[0].conditions === newConditionsValue).toBeTruthy();
  });
});

describe('setAddNewMode', () => {
  it('should set addNewMode when type is value', () => {
    let newRules: TRule[] = [];
    const newAddNewModeValue = false;
    const oldRules: Partial<TRule>[] = [
      {
        addNewMode: true,
      },
    ];
    const setRules = (newValue: (oldRules: TRule[]) => TRule[]) => {
      newRules = newValue(oldRules as TRule[]);
    };
    const setAddNewModeFn = setAddNewMode(setRules as IRules['setRules'], 0);
    setAddNewModeFn(newAddNewModeValue);
    expect(newRules).not.toBeUndefined();
    expect(newRules[0].addNewMode).toEqual(newAddNewModeValue);
  });

  it('should set addNewMode when type is function', () => {
    let newRules: TRule[] = [];
    const newAddNewModeValueFn = () => false;
    const oldRules: Partial<TRule>[] = [
      {
        addNewMode: true,
      },
    ];
    const setRules = (newValue: (oldRules: TRule[]) => TRule[]) => {
      newRules = newValue(oldRules as TRule[]);
    };
    const setAddNewModeFn = setAddNewMode(setRules as IRules['setRules'], 0);
    setAddNewModeFn(newAddNewModeValueFn);
    expect(newRules).not.toBeUndefined();
    expect(newRules[0].addNewMode).toEqual(newAddNewModeValueFn());
  });
});
