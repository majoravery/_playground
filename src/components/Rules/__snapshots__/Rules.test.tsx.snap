// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rules component should be able to add and remove rules and conditions 1`] = `
NodeList [
  <div
    class="view-condition__container"
    data-testid="view-condition-container"
  >
    <div
      class="dropdown--component dropdown"
    >
      <div>
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="button--component dropdown-toggle btn btn-outline-primary"
          id="key-1-logical-operator"
          type="button"
        >
          <span>
            If
          </span>
        </button>
      </div>
    </div>
    <div
      class="rule-label__container"
      style="width: 184px;"
    >
      New Title
    </div>
    <div
      class="rule-label__container"
      style="width: 72px;"
    >
      is
    </div>
    <div
      class="values__container"
    >
      <div
        class="values__content-container"
      >
        <div
          style="margin-top: -1px; margin-bottom: -1px;"
        >
          <div>
            neu1
          </div>
          <div>
            new2
          </div>
        </div>
      </div>
      <div
        class="values__button-container"
      >
        <button
          class="button--component button--component--icon-only btn btn-light btn-sm"
          data-testid="rule-edit-icon"
          type="button"
        >
          <svg
            fill="none"
            height="14"
            opacity="1"
            stroke="#051022"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 4H4a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-7"
            />
            <path
              d="M18.5 2.5a2.121 2.121 0 0 1 3 3L12 15l-4 1 1-4 9.5-9.5z"
            />
          </svg>
        </button>
        <button
          class="button--component button--component--icon-only btn btn-light btn-sm"
          data-testid="rule-delete-icon"
          type="button"
        >
          <svg
            fill="none"
            height="14"
            opacity="1"
            stroke="#051022"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="3 6 5 6 21 6"
            />
            <path
              d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
            />
            <line
              x1="10"
              x2="10"
              y1="11"
              y2="17"
            />
            <line
              x1="14"
              x2="14"
              y1="11"
              y2="17"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>,
  <div
    class="view-condition__container"
    data-testid="view-condition-container"
  >
    <div
      class="dropdown--component dropdown"
    >
      <div>
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="button--component dropdown-toggle btn btn-outline-primary"
          id="key-2-logical-operator"
          type="button"
        >
          <span>
            And
          </span>
        </button>
      </div>
    </div>
    <div
      class="rule-label__container"
      style="width: 184px;"
    >
      New Title
    </div>
    <div
      class="rule-label__container"
      style="width: 72px;"
    >
      is
    </div>
    <div
      class="values__container"
    >
      <div
        class="values__content-container"
      >
        <div
          style="margin-top: -1px; margin-bottom: -1px;"
        >
          <div>
            neu1
          </div>
          <div>
            new2
          </div>
        </div>
      </div>
      <div
        class="values__button-container"
      >
        <button
          class="button--component button--component--icon-only btn btn-light btn-sm"
          data-testid="rule-edit-icon"
          type="button"
        >
          <svg
            fill="none"
            height="14"
            opacity="1"
            stroke="#051022"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 4H4a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-7"
            />
            <path
              d="M18.5 2.5a2.121 2.121 0 0 1 3 3L12 15l-4 1 1-4 9.5-9.5z"
            />
          </svg>
        </button>
        <button
          class="button--component button--component--icon-only btn btn-light btn-sm"
          data-testid="rule-delete-icon"
          type="button"
        >
          <svg
            fill="none"
            height="14"
            opacity="1"
            stroke="#051022"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="3 6 5 6 21 6"
            />
            <path
              d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
            />
            <line
              x1="10"
              x2="10"
              y1="11"
              y2="17"
            />
            <line
              x1="14"
              x2="14"
              y1="11"
              y2="17"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>,
  <div
    class="view-condition__container"
    data-testid="view-condition-container"
  >
    <div
      class="dropdown--component dropdown"
    >
      <div>
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="button--component dropdown-toggle btn btn-outline-primary"
          id="key-3-logical-operator"
          type="button"
        >
          <span>
            And
          </span>
        </button>
      </div>
    </div>
    <div
      class="rule-label__container"
      style="width: 184px;"
    >
      New Title
    </div>
    <div
      class="rule-label__container"
      style="width: 72px;"
    >
      is
    </div>
    <div
      class="values__container"
    >
      <div
        class="values__content-container"
      >
        <div
          style="margin-top: -1px; margin-bottom: -1px;"
        >
          <div>
            neu1
          </div>
          <div>
            new2
          </div>
        </div>
      </div>
      <div
        class="values__button-container"
      >
        <button
          class="button--component button--component--icon-only btn btn-light btn-sm"
          data-testid="rule-edit-icon"
          type="button"
        >
          <svg
            fill="none"
            height="14"
            opacity="1"
            stroke="#051022"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 4H4a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-7"
            />
            <path
              d="M18.5 2.5a2.121 2.121 0 0 1 3 3L12 15l-4 1 1-4 9.5-9.5z"
            />
          </svg>
        </button>
        <button
          class="button--component button--component--icon-only btn btn-light btn-sm"
          data-testid="rule-delete-icon"
          type="button"
        >
          <svg
            fill="none"
            height="14"
            opacity="1"
            stroke="#051022"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="3 6 5 6 21 6"
            />
            <path
              d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
            />
            <line
              x1="10"
              x2="10"
              y1="11"
              y2="17"
            />
            <line
              x1="14"
              x2="14"
              y1="11"
              y2="17"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>,
]
`;

exports[`Rules component should be able to add and remove rules and conditions 2`] = `
<div
  class="view-condition__container"
  data-testid="view-condition-container"
>
  <div
    class="rule-label__container"
    style="width: 112px;"
  >
    If
  </div>
  <div
    class="rule-label__container"
    style="width: 184px;"
  >
    New Title
  </div>
  <div
    class="rule-label__container"
    style="width: 72px;"
  >
    is
  </div>
  <div
    class="values__container"
  >
    <div
      class="values__content-container"
    >
      <div
        style="margin-top: -1px; margin-bottom: -1px;"
      >
        <div>
          neu1
        </div>
        <div>
          new2
        </div>
      </div>
    </div>
    <div
      class="values__button-container"
    >
      <button
        class="button--component button--component--icon-only btn btn-light btn-sm"
        type="button"
      >
        <svg
          fill="none"
          height="14"
          opacity="1"
          stroke="#051022"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1 12s4-8 11-8 11 8 11 8-4 8-11 8-11-8-11-8z"
          />
          <circle
            cx="12"
            cy="12"
            r="3"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`Rules component should match snapshot when empty 1`] = `
<div>
  <div
    class="rules__container"
    data-testid="rules--container"
  />
</div>
`;

exports[`Rules component should match snapshot when empty 2`] = `
<div>
  <div
    class="rules__container"
    data-testid="rules--container"
  >
    <button
      class="button--component rule__add-button btn btn-outline-primary"
      data-testid="rules--add-rule-button"
      type="button"
    >
      <svg
        fill="none"
        height="14"
        opacity="1"
        stroke="#051022"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="14"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="12"
          cy="12"
          r="10"
        />
        <line
          x1="12"
          x2="12"
          y1="8"
          y2="16"
        />
        <line
          x1="8"
          x2="16"
          y1="12"
          y2="12"
        />
      </svg>
      <span
        class="text-with-icons"
      >
        Add rule
      </span>
    </button>
  </div>
</div>
`;
